Tools/Scripts/hooks/pre-push

```diff
@@ -0,0 +1,330 @@
+    if not name_to_remote:
+        _, name_to_remote = remote_name_mappings()
+
+    # Class 1: the commit we're pushing exists on a remote already
+    if remotes:
+        return security_level_for_remotes(remotes, name_to_remote), CLASS_1
+
+    message = message_for(commit)
+    title = message.splitlines()[0]
+
+    # Class 2: The commit we're pushing is a cherry-pick of something that exists on a remote already
+    for r in CHERRY_PICK_RE:
+        match = r.match(title)
+        if not match:
+            continue
+        primary = match.group('primary')
+        secondary = match.group('secondary')
+        if secondary:
+            secondary = UNPACK_SECONDARY_RE.match(secondary).groups()[0]
+        for candidate in [primary, secondary]:
+            original_remotes = remotes_for(candidate)
+            if not original_remotes:
+                continue
+            return security_level_for_remotes(original_remotes, name_to_remote), CLASS_2
+        break
+
+    # Class 3: We must inspect the commit for bug references
+    if REPOSITORY:
+        obj = Commit(hash=commit, message=message)
+        for issue in obj.issues:
+            if issue.redacted:
+                return 1, CLASS_3
+        return 0, CLASS_3
+
+    return None, None
+
+
+def main(name, remote):
+    target_security_level = security_level_for_remote(name, remote)
+    if VERBOSITY > QUIET:
+        if target_security_level:
+            print('Pushing to {}, which is classified as a secure remote'.format(name or remote))
+        elif target_security_level == 0:
+            print('Pushing to {}, which is classified as a public remote'.format(name or remote))
+        else:
+            print('Pushing to {}, which has not classification. Treating it as a public remote'.format(name or remote))
+
+    to_push = []
+    for push in sys.stdin.readlines():
+        _, local_ref, _, remote_ref = push.split()
+        # Deletes are always safe
+        if local_ref == NULL_HASH:
+            continue
+        if remote_ref == NULL_HASH:
+            remote_ref = None
+        to_push.append((remote_ref, local_ref))
+    if not to_push:
+        if VERBOSITY > QUIET:
+            print('No novel content being pushed')
+        return 0
+
+    remote_to_name, name_to_remote = remote_name_mappings()
+    if name == remote:
+        match = REMOTE_RE.match(remote)
+        if not match:
+            print_error("'{}' does not conform to a recognized target remote".format(remote))
+            return 255
+        name = remote_to_name.get('{}:{}'.format(match.group('host'), match.group('path')), None)
+        if not name:
+            print_error("Failed to map '{}' to an existing remote".format(remote))
+            return 1
+
+
+    remotes_with = {}
+    for from_ref, to_ref in to_push:
+        for commit in commits_in(from_ref, to_ref):
+            if commit not in remotes_with:
+                remotes_with[commit] = remotes_for(commit) or []
```

Comment:
This is 1 `git branch` call per commit, right? That strikes me as unacceptably slow. For instance if I do:

```
git fetch origin main:main
git push emw main:main
```

that is potentially pushing thousands of commits.

Naively, I would think we only need to checks which remotes have `from_ref`, because if they have any commit in `from_ref+1..to_ref`, they have from_ref too.

Reply:
`from_ref` will often exist on many more remotes than `to_ref`...actually, I suppose it's the commits closer to `to_ref` which are, generally, more interesting.

We could probably speed this up with an early exit, I think. The first time we encounter a commit which exists on a classifiable remote, that's probably grounds for an early exit.